Add support for tuples

davesque opened the pull request:
### What was wrong?

This is a rebuild of PR #1147 .  Commits retain authorship from original PR where appropriate.

### How was it fixed?

Adds support for tuples by modifying low-level utility functions used for ABI dict processing and encodability checks.

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://pixfeeds.com/images/animals/red-pandas/1280-178167661-red-panda.jpg)

davesque commented:
I'm adding things bit by bit here.  The most recent chunk of work done was to rewrite the `get_abi_inputs` utility function to be simpler and more robust.

Eventually, I'll give things another once over to make sure I didn't miss anything.

Known issues:
* Updated `get_abi_*_types` functions need to have updated tests
* New `get_abi_inputs` function needs more tests.  Will probably add some or all of the tests written by @feuGeneA for that function.

pipermerriam commented on web3/_utils/abi.py:
```diff
@@ -342,6 +347,66 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
+    # Component is tuple (possibly tuple list)...
+
+    array_dims = abi_type_str[TUPLE_PREFIX_LEN:]
+    if array_dims:
```
Can we tighten this up at all?  Match against a regex or something?  This jumps out as a good way to end up with weird errors jumping out in the case that the ABI type string is malformed.

davesque replied:
Good idea.  I'll add that.

pipermerriam commented on web3/_utils/abi.py:
```diff
@@ -342,6 +347,66 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
+        new_comp = copy.copy(comp)
+        new_comp['type'] = TUPLE_PREFIX
+
+        # Assume `arg` is iterable
```
Same here, easy to do a quick runtime check that `isinstance(arg, collections.abc.Iterable)` or maybe tighter as `not isinstance(arg, (bytes, str)) and isinstance(arg, collections.abc.Sequence)`

davesque replied:
Added a runtime type check using `eth_utils.is_list_like`.

pipermerriam commented on web3/_utils/abi.py:
```diff
@@ -342,6 +347,66 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
+
+    if isinstance(arg, abc.Mapping):
+        # Arg is mapping.  Convert to properly ordered sequence.
+        arg = tuple(arg[c['name']] for c in sub_comps)
```
It's going to take me a long time to not have a knee-jerk reflex that python dictionaries aren't ordered (since they now are)

davesque replied:
If you meant `sub_comps`, I believe that should always be a `list` instance.

carver commented on web3/_utils/abi.py:
```diff
@@ -302,6 +313,85 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
+
+    if isinstance(args, abc.Mapping):
+        # `args` is mapping, convert to properly ordered sequence
+        args = tuple(args[i['name']] for i in inputs)
```
```suggestion
        args = tuple(args[input['name']] for input in inputs)
```

*nit* opinion: even in short comprehensions, I think it's easier to read with a full variable name. It prevents your eyes having to dart back and forth, so you can more or less read it straight through in one pass.

carver commented on web3/_utils/abi.py:
```diff
@@ -302,6 +313,85 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
+
+def _convert_abi_input(comp, arg):
+    """
+    Converts an argument ``arg`` corresponding to an ABI component ``comp``
```
`component` here seems confusing because each element of a tuple is called `component`. Maybe this is a `arg_type`, or `arg_spec` since it includes a name?

carver commented on tests/core/contracts/test_contract_call_interface.py:
```diff
@@ -612,3 +617,61 @@ def test_invalid_fixed_value_reflections(web3, fixed_reflection_contract, functi
+            ]),
+        ),
+    ),
+)
```
It looks like this is already working, but might be nice to add a regression test for nested arrays of tuples. Seems like the kind of thing that could break in a rewrite/refactor.

carver commented on web3/_utils/abi.py:
```diff
@@ -580,11 +670,25 @@ def abi_sub_tree(abi_type: Optional[Union[TypeStr, ABIType]], data_value: Any) -
+        # If type is array, determine item type and annotate all
+        # items in iterable with that type
+        item_type_str = abi_type.item_type.to_type_str()
+        data_value = [abi_sub_tree(item_type_str, i) for i in data_value]
```
Maybe `item` instead of `i`? Not sure if that's what `i` is short for, here.

carver commented on web3/_utils/abi.py:
```diff
@@ -580,11 +670,25 @@ def abi_sub_tree(abi_type: Optional[Union[TypeStr, ABIType]], data_value: Any) -
+        # items in iterable respectively with those types
+        data_value = type(data_value)(
+            abi_sub_tree(comp_type.to_type_str(), i)
+            for comp_type, i in zip(abi_type.components, data_value)
```
Maybe `component` instead of `i` here?

carver commented on web3/_utils/contracts.py:
```diff
@@ -240,6 +241,8 @@ def get_function_info(fn_name, contract_abi=None, fn_abi=None, args=None, kwargs
 
     fn_arguments = merge_args_and_kwargs(fn_abi, args, kwargs)
 
+    _, fn_arguments = get_abi_inputs(fn_abi, fn_arguments)
```
*nit* overwrote `fn_arguments`, but now it has a slightly different meaning. Maybe `typed_arguments` or `annotated_arguments`? (same as the one in `web3/_utils/abi` I commented on)

davesque commented on tests/core/contracts/conftest.py:
```diff
@@ -6,6 +6,53 @@
+"""
+CONTRACT_TUPLE_CODE = '0x608060405234801561001057600080fd5b506108ca806100206000396000f300608060405260043610610041576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680638e1ae3c714610046575b600080fd5b34801561005257600080fd5b5061006d60048036036100689190810190610403565b610083565b60405161007a9190610696565b60405180910390f35b61008b610093565b819050919050565b60606040519081 ...
+CONTRACT_TUPLE_RUNTIME = '0x608060405260043610610041576000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff1680638e1ae3c714610046575b600080fd5b34801561005257600080fd5b5061006d60048036036100689190810190610403565b610083565b60405161007a9190610696565b60405180910390f35b61008b610093565b819050919050565b6060604051908101604052806000815260200160608152602001606081525090565b6000610 ...
+CONTRACT_TUPLE_ABI = json.loads('[{"constant":true,"inputs":[{"components":[{"name":"a","type":"uint256"},{"name":"b","type":"uint256[]"},{"components":[{"name":"x","type":"int256"},{"name":"y","type":"bool[2]"},{"name":"z","type":"address[]"}],"name":"c","type":"tuple[]"}],"name":"s","type":"tuple"}],"name":"method","outputs":[{"components":[{"name":"a","type":"uint256"},{"name":"b","type":"uint ...
```
I was just following the style of existing fixture code with that one.

davesque commented on tests/core/contracts/test_contract_method_to_argument_matching.py:
```diff
@@ -13,8 +13,82 @@
 SINGLE_FN_ONE_ARG = json.loads('[{"constant":false,"inputs":[{"name":"","type":"uint256"}],"name":"a","outputs":[],"type":"function"}]')  # noqa: E501
-MULTIPLE_FUNCTIONS = json.loads('[{"constant":false,"inputs":[],"name":"a","outputs":[],"type":"function"},{"constant":false,"inputs":[{"name":"","type":"bytes32"}],"name":"a","outputs":[],"type":"function"},{"constant":false,"inputs":[{"name":"","type":"uint256"}],"name":"a","outputs":[],"type":"function"},{"constant":false,"inputs":[{"name":"","type":"uint8"}],"name":"a","outputs":[],"type":"fu ...
 FALLBACK_FUNCTION = json.loads('[{"constant": false, "inputs": [], "name": "getData", "outputs": [{"name": "r", "type": "uint256"}], "payable": false, "stateMutability": "nonpayable", "type": "function"}, {"payable": false, "stateMutability": "nonpayable", "type": "fallback"}]')  # noqa: E501
+MULTIPLE_FUNCTIONS = json.loads('''
```
Was following the style of existing code.

davesque commented:
@carver I sort of feel like avoiding single letter variables can be just as confusing.  In a lot of the examples cited, the cognitive overhead of the short variable names seems pretty minimal.  Also, giving things longer names often leads to a code section taking up two or three lines when it could have taken up one.  I feel like the overhead of reading a statement that runs across multiple lines is as much if not more than the overhead of reading a statement that uses meaningful abbreviations.  Additionally, the typical alternative of using a variable name that simply lacks pluralization is pretty confusing because then two variables only differ by one letter and it's easy to get typos.

Anyway, I'll change it if we really want but I've often felt like I don't understand the aversion to that coding pattern.  I wanted to explain some of my thoughts about that.

davesque commented:
@carver I went ahead and addressed all of your feedback.  As far as the testing of deeper tuple nesting goes, is [this](https://github.com/ethereum/web3.py/pull/1235/commits/140e32eca501830aa59c63d29d788bb3b0476efb) what you had in mind?

carver commented:
> As far as the testing of deeper tuple nesting goes, is [this](https://github.com/ethereum/web3.py/pull/1235/commits/140e32eca501830aa59c63d29d788bb3b0476efb) what you had in mind?

Yup, that's about it @davesque ! :) I suspect there are maybe funkier things that could go wrong if someone messes about with this code, but I think this is plenty of coverage for now. :+1: :shipit:

pipermerriam commented:
@davesque can you confirm this is :+1: to merge (i.e. you don't have anything outstanding that you know needs to be addressed).

@kclowes or maybe @njgheorghita .  IIUC correctly, this gives us support for tuples, but only as raw tuples, and not any *rich* object representations.

I think what comes next is some version of support for those.

1. named tuples
2. dictionaries
3. arbitrary objects with attributes.

It isn't clear to me exactly how these should be implemented but I think doing them in this order probably makes sense.  The `NamedTuple` one is probably easiest since we can generate the class from the ABI definition and return it in place of a plain tuple.  This could be a core part of how web3.py extends the eth-abi API.

davesque commented:
@pipermerriam Yes it's ready to merge.  I was waiting for your review.

pipermerriam commented on web3/_utils/abi.py:
```diff
@@ -302,6 +313,90 @@ def merge_args_and_kwargs(function_abi, args, kwargs):
+
+    if isinstance(arg, abc.Mapping):
+        # Arg is mapping.  Align values according to abi order.
+        arg = tuple(arg[abi['name']] for abi in sub_abis)
```
We have a convention of avoiding overwriting local variables, preferring to instead use new variable names.  Maybe we can call this `aligned_arg`

```python
if is_mapping(arg):
    aligned_arg = tuple(...)
else:
    aligned_arg = arg
```

> NOTE: the above statement has an implicit `else` clause which is also somewhat against our conventions.  It might should look something like.

```python
if is_mapping(arg):
    ...
elif is_list_like(arg):
    ...
else:
    raise TypeError("Something something invalid")
```

This would simply be moving the `TypeError` from below up to be part of this clause.

pipermerriam replied:
Also noting this relates to #1267 as it appears we already support dictionaries for the input arguments for tuple types.

pipermerriam commented on web3/_utils/abi.py:
```diff
@@ -580,11 +675,28 @@ def abi_sub_tree(abi_type: Optional[Union[TypeStr, ABIType]], data_value: Any) -
+        # If type is array, determine item type and annotate all
+        # items in iterable with that type
+        item_type_str = abi_type.item_type.to_type_str()
+        data_value = [
```
Same here with convention of not overwriting local variables which if fixed will require this `if/elif` to get a final `else` clause.

davesque commented:
@carver @pipermerriam Thanks guys for the reviews.  When the tests pass, I'll go ahead and merge this.
